README.md
# bmkg_gempa

[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT) 
[![pub package](https://img.shields.io/pub/v/bmkg_gempa.svg)](https://pub.dartlang.org/packages/bmkg_gempa)

Dart client library untuk mendapat data gempabumi dari situs data terbuka BMKG.

## Contoh Penggunaan

```dart
import 'package:bmkg_gempa/bmkg_gempa.dart';

void main() async {
  // init client BmkgGempa
  var client = BmkgGempa();

  // mendapatkan satu data gempa terbaru
  Gempa gempaTerbaru = await client.gempaTerbaru();
  print('--------------------------------');
  print('Gempabumi terbaru:');
  print("Tanggal : ${gempaTerbaru.tanggal}");
  print("Jam : ${gempaTerbaru.jam}");
  print("DateTime : ${gempaTerbaru.dateTime}");
  print("Coordinates : ${gempaTerbaru.coordinates}");
  print("Lintang : ${gempaTerbaru.lintang}");
  print("Bujur : ${gempaTerbaru.bujur}");
  print("Magnitude : ${gempaTerbaru.magnitude}");
  print("Kedalaman : ${gempaTerbaru.kedalaman}");
  print("Wilayah : ${gempaTerbaru.wilayah}");
  print("Potensi : ${gempaTerbaru.potensi}");
  print("Dirasakan : ${gempaTerbaru.dirasakan}");
  print("Shakemap : ${gempaTerbaru.shakemap}");
  print('--------------------------------');

  // mendapatkan kumpulan data gempa terkini
  List<Gempa> gempaTerkini = await client.gempaTerkini();
  print('Daftar 15 Gempabumi M 5.0+:');
  for (var gempa in gempaTerkini) {
    print('--------------------------------');
    print("Tanggal : ${gempa.tanggal}");
    print("Jam : ${gempa.jam}");
    print("DateTime : ${gempa.dateTime}");
    print("Coordinates : ${gempa.coordinates}");
    print("Lintang : ${gempa.lintang}");
    print("Bujur : ${gempa.bujur}");
    print("Magnitude : ${gempa.magnitude}");
    print("Kedalaman : ${gempa.kedalaman}");
    print("Wilayah : ${gempa.wilayah}");
    print("Potensi : ${gempa.potensi}");
    print('--------------------------------');
  }

  // mendapatkan kumpulan data gempa yang dirasakan
  List<Gempa> gempaDirasakan = await client.gempaDirasakan();
  print('Daftar 15 Gempabumi Dirasakan: ');
  for (var gempa in gempaDirasakan) {
    print('--------------------------------');
    print("Tanggal : ${gempa.tanggal}");
    print("Jam : ${gempa.jam}");
    print("DateTime : ${gempa.dateTime}");
    print("Coordinates : ${gempa.coordinates}");
    print("Lintang : ${gempa.lintang}");
    print("Bujur : ${gempa.bujur}");
    print("Magnitude : ${gempa.magnitude}");
    print("Kedalaman : ${gempa.kedalaman}");
    print("Wilayah : ${gempa.wilayah}");
    print("Dirasakan : ${gempa.dirasakan}");
    print('--------------------------------');
  }
}
```

## License

MIT

## Credits

Sumber data pada package ini berasal dari situs Data Terbuka milik BMKG(Badan Meteorologi, Klimatologi, dan Geofisika) - [Link](https://data.bmkg.go.id/gempabumi/)

CHANGELOG.md
## 1.0.2

* Remove flutter dependencies

## 1.0.1

* Add API Testing

## 1.0.0+1

* Update API doc and code formatting

## 1.0.0

* Initial release


test/bmkg_gempa_test.dart
import 'package:bmkg_gempa/bmkg_gempa.dart';
import 'package:test/test.dart';

void main() {
  late BmkgGempa client;

  setUpAll(() {
    client = BmkgGempa();
  });
  test('mendapatkan satu data gempa terbaru', () async {
    var gempaTerbaru = await client.gempaTerbaru();
    expect(gempaTerbaru, isA<Gempa>());
    expect(gempaTerbaru.bujur, isA<String>());
    expect(gempaTerbaru.lintang, isA<String>());
    expect(gempaTerbaru.coordinates, isA<String>());
    expect(gempaTerbaru.dateTime, isA<DateTime>());
    expect(gempaTerbaru.dirasakan, isA<String>());
    expect(gempaTerbaru.jam, isA<String>());
    expect(gempaTerbaru.kedalaman, isA<String>());
    expect(gempaTerbaru.magnitude, isA<String>());
    expect(gempaTerbaru.potensi, isA<String>());
    expect(gempaTerbaru.shakemap, isA<String>());
    expect(gempaTerbaru.tanggal, isA<String>());
    expect(gempaTerbaru.wilayah, isA<String>());
  });

  test('mendapatkan kumpulan data gempa terkini', () async {
    var gempaList = await client.gempaTerkini();
    var gempa = gempaList[0];
    expect(gempaList, isA<List<Gempa>>());
    expect(gempaList.length, 15);
    expect(gempa, isA<Gempa>());
    expect(gempa.bujur, isA<String>());
    expect(gempa.lintang, isA<String>());
    expect(gempa.coordinates, isA<String>());
    expect(gempa.dateTime, isA<DateTime>());
    expect(gempa.dirasakan, null);
    expect(gempa.jam, isA<String>());
    expect(gempa.kedalaman, isA<String>());
    expect(gempa.magnitude, isA<String>());
    expect(gempa.potensi, isA<String>());
    expect(gempa.shakemap, null);
    expect(gempa.tanggal, isA<String>());
    expect(gempa.wilayah, isA<String>());
  });

  test('mendapatkan kumpulan data gempa yang dirasakan', () async {
    var gempaList = await client.gempaDirasakan();
    var gempa = gempaList[0];
    expect(gempaList, isA<List<Gempa>>());
    expect(gempaList.length, 15);
    expect(gempa, isA<Gempa>());
    expect(gempa.bujur, isA<String>());
    expect(gempa.lintang, isA<String>());
    expect(gempa.coordinates, isA<String>());
    expect(gempa.dateTime, isA<DateTime>());
    expect(gempa.dirasakan, isA<String>());
    expect(gempa.jam, isA<String>());
    expect(gempa.kedalaman, isA<String>());
    expect(gempa.magnitude, isA<String>());
    expect(gempa.potensi, null);
    expect(gempa.shakemap, null);
    expect(gempa.tanggal, isA<String>());
    expect(gempa.wilayah, isA<String>());
  });
}


test/fixture/fixture_reader.dart
import 'dart:io';

String fixture(String name) {
  var currentDirectory = Directory.current.toString().replaceAll('\'', '');
  var lastDirectory = currentDirectory.split('/')[currentDirectory.split('/').length - 1];
  if (lastDirectory == 'test') {
    return File('fixture/$name').readAsStringSync();
  } else {
    return File('test/fixture/$name').readAsStringSync();
  }
}


lib/src/bmkg_gempa.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:bmkg_gempa/src/models/gempa.dart';

/// Data kejadian gempabumi yang terjadi di seluruh wilayah Indonesia.
/// Terdapat 3 jenis data kejadian gempabumi, yaitu :
/// Gempabumi M 5.0+, Gempabumi Dirasakan, dan Gempabumi Terbaru
class BmkgGempa {
  final String _baseUrl = 'data.bmkg.go.id';
  final String _path = '/DataMKG/TEWS/';

  Future<dynamic> _getRequest(String url, {Map<String, String>? query}) async {
    var uri = Uri.https(_baseUrl, url, query);
    var response = await http.get(uri);
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Request failed with status: ${response.statusCode}.');
    }
  }

  /// Mendapatkan data gempabumi terbaru
  ///
  /// Mengembalikan object [Gempa] jika pemanggilan sukses
  /// Nilai properti yang tersedia pada method ini berupa :
  /// - Tanggal
  /// - Jam
  /// - DateTime
  /// - Coordinates
  /// - Lintang
  /// - Bujur
  /// - Magnitude
  /// - Kedalaman
  /// - Wilayah
  /// - Potensi
  /// - Dirasakan
  /// - Shakemap
  ///
  /// Throws [Exception] jika terjadi error
  Future<Gempa> gempaTerbaru() async {
    var url = '${_path}autogempa.json';
    var response = await _getRequest(url);
    var json = response['Infogempa']['gempa'];
    return Gempa.fromJson(json);
  }

  /// Mendapatkan daftar 15 data gempabumi M 5.0+
  ///
  /// Mengembalikan object List [Gempa] jika pemanggilan sukses
  /// Nilai properti yang tersedia pada method ini berupa :
  /// - Tanggal
  /// - Jam
  /// - DateTime
  /// - Coordinates
  /// - Lintang
  /// - Bujur
  /// - Magnitude
  /// - Kedalaman
  /// - Wilayah
  /// - Potensi
  ///
  /// Throws [Exception] jika terjadi error
  Future<List<Gempa>> gempaTerkini() async {
    var url = '${_path}gempaterkini.json';
    var response = await _getRequest(url);

    var gempaList = <Gempa>[];
    List json = response['Infogempa']['gempa'];
    for (var e in json) {
      var gempa = Gempa.fromJson(e);
      gempaList.add(gempa);
    }
    return gempaList;
  }

  /// Mendapatkan daftar 15 data gempabumi yang dirasakan
  ///
  /// Mengembalikan object List [Gempa] jika pemanggilan sukses.
  /// Nilai properti yang tersedia pada method ini berupa :
  /// - Tanggal
  /// - Jam
  /// - DateTime
  /// - Coordinates
  /// - Lintang
  /// - Bujur
  /// - Magnitude
  /// - Kedalaman
  /// - Wilayah
  /// - Dirasakan
  ///
  /// Throws [Exception] jika terjadi error
  Future<List<Gempa>> gempaDirasakan() async {
    var url = '${_path}gempadirasakan.json';
    var response = await _getRequest(url);

    var gempaList = <Gempa>[];
    List json = response['Infogempa']['gempa'];
    for (var e in json) {
      var gempa = Gempa.fromJson(e);
      gempaList.add(gempa);
    }
    return gempaList;
  }
}


lib/src/models/gempa.dart
/// Model class [Gempa] digunakan untuk parsing data ketika melakukan
/// pemanggilan method untuk mendapatkan data gempabumi.
class Gempa {
  static final String _baseUrl = 'https://data.bmkg.go.id/DataMKG/TEWS/';

  /// [tanggal] dalam bentuk WIB
  final String? tanggal;

  /// [jam] dalam bentuk WIB
  final String? jam;

  /// [dateTime] sesuai dalam UTC (+00:00)
  final DateTime? dateTime;

  /// [coordinates] yang merupakan gabungan latitude kemudian longitude
  final String? coordinates;

  /// koordinat [lintang] dari lokasi gempa
  final String? lintang;

  /// koordinat [bujur] dari lokasi gempa
  final String? bujur;

  /// [magnitude] merupakan kekuatan gempa
  final String? magnitude;

  /// [kedalaman] kedalaman gempa dalam satuan kilometer(km)
  final String? kedalaman;

  /// [wilayah] terdekat dengan lokasi episenter gempabumi
  final String? wilayah;

  /// [potensi] tsunami atau tidak, dan status gempa dirasakan
  final String? potensi;

  /// [dirasakan] merupakan wilayah yang merasakan gempa dalam skala MMI
  final String? dirasakan;

  /// Gambar [shakemap] (peta guncangan) diawali dengan URL `https://data.bmkg.go.id/DataMKG/TEWS/`
  final String? shakemap;

  Gempa({
    this.tanggal,
    this.jam,
    this.dateTime,
    this.coordinates,
    this.lintang,
    this.bujur,
    this.magnitude,
    this.kedalaman,
    this.wilayah,
    this.potensi,
    this.dirasakan,
    this.shakemap,
  });

  /// Parsing data [json] menjadi objek [Gempa]
  factory Gempa.fromJson(Map<String, dynamic> json) => Gempa(
        tanggal: json['Tanggal'] as String?,
        jam: json['Jam'] as String?,
        dateTime: json['DateTime'] == null
            ? null
            : DateTime.parse(json['DateTime'] as String),
        coordinates: json['Coordinates'] as String?,
        lintang: json['Lintang'] as String?,
        bujur: json['Bujur'] as String?,
        magnitude: json['Magnitude'] as String?,
        kedalaman: json['Kedalaman'] as String?,
        wilayah: json['Wilayah'] as String?,
        potensi: json['Potensi'] as String?,
        dirasakan: json['Dirasakan'] as String?,
        shakemap: json['Shakemap'] == null
            ? null
            : '$_baseUrl${(json['Shakemap'] as String?)}',
      );
}


example/main.dart
import 'package:bmkg_gempa/bmkg_gempa.dart';

void main() async {
  // init client BmkgGempa
  var client = BmkgGempa();

  // mendapatkan satu data gempa terbaru
  Gempa gempaTerbaru = await client.gempaTerbaru();
  print('--------------------------------');
  print('Gempabumi terbaru:');
  print("Tanggal : ${gempaTerbaru.tanggal}");
  print("Jam : ${gempaTerbaru.jam}");
  print("DateTime : ${gempaTerbaru.dateTime}");
  print("Coordinates : ${gempaTerbaru.coordinates}");
  print("Lintang : ${gempaTerbaru.lintang}");
  print("Bujur : ${gempaTerbaru.bujur}");
  print("Magnitude : ${gempaTerbaru.magnitude}");
  print("Kedalaman : ${gempaTerbaru.kedalaman}");
  print("Wilayah : ${gempaTerbaru.wilayah}");
  print("Potensi : ${gempaTerbaru.potensi}");
  print("Dirasakan : ${gempaTerbaru.dirasakan}");
  print("Shakemap : ${gempaTerbaru.shakemap}");
  print('--------------------------------');

  // mendapatkan kumpulan data gempa terkini
  List<Gempa> gempaTerkini = await client.gempaTerkini();
  print('Daftar 15 Gempabumi M 5.0+:');
  for (var gempa in gempaTerkini) {
    print('--------------------------------');
    print("Tanggal : ${gempa.tanggal}");
    print("Jam : ${gempa.jam}");
    print("DateTime : ${gempa.dateTime}");
    print("Coordinates : ${gempa.coordinates}");
    print("Lintang : ${gempa.lintang}");
    print("Bujur : ${gempa.bujur}");
    print("Magnitude : ${gempa.magnitude}");
    print("Kedalaman : ${gempa.kedalaman}");
    print("Wilayah : ${gempa.wilayah}");
    print("Potensi : ${gempa.potensi}");
    print('--------------------------------');
  }

  // mendapatkan kumpulan data gempa yang dirasakan
  List<Gempa> gempaDirasakan = await client.gempaDirasakan();
  print('Daftar 15 Gempabumi Dirasakan: ');
  for (var gempa in gempaDirasakan) {
    print('--------------------------------');
    print("Tanggal : ${gempa.tanggal}");
    print("Jam : ${gempa.jam}");
    print("DateTime : ${gempa.dateTime}");
    print("Coordinates : ${gempa.coordinates}");
    print("Lintang : ${gempa.lintang}");
    print("Bujur : ${gempa.bujur}");
    print("Magnitude : ${gempa.magnitude}");
    print("Kedalaman : ${gempa.kedalaman}");
    print("Wilayah : ${gempa.wilayah}");
    print("Dirasakan : ${gempa.dirasakan}");
    print('--------------------------------');
  }
}
